crates/binjs_es6/src/sublanguages.rs

```diff
@@ -0,0 +1,426 @@
+            // If either the parent or the child is complex, the parent is complex.
+            (&Properties::Complex, _) | (_, &Properties::Complex) => Properties::Complex,
+
+            // If both are pure so far, add both sizes.
+            (
+                &Properties::PureDataSoFar { size: my_size },
+                &Properties::PureDataSoFar { size: parent_size },
+            ) => Properties::PureDataSoFar {
+                size: my_size + parent_size,
+            },
+        };
+        *parent = new_parent
+    }
+}
+
+impl WalkGuard<InjectVisitor> for PropertiesGuard {
+    /// Constructor called by default methods of `InjectVisitor`
+    fn new(origin: &InjectVisitor, path: &WalkPath) -> Self {
+        PropertiesGuard::complex(origin, path)
+    }
+}
+
+type EnterResult = Result<VisitMe<PropertiesGuard>, ()>;
+type ExitResult<T> = Result<Option<T>, ()>;
+
+impl InjectVisitor {
+    pub fn new(pure_data_threshold: usize) -> Self {
+        Self {
+            pure_data_threshold,
+            stack: Rc::new(RefCell::new(Vec::new())),
+        }
+    }
+    pub fn rewrite_script(pure_data_threshold: usize, script: &mut Script) {
+        let mut visitor = Self::new(pure_data_threshold);
+        script
+            .walk(&mut WalkPath::new(), &mut visitor)
+            .expect("Could not walk script");
+    }
+
+    /// The properties of the current node.
+    fn properties(&self) -> Properties {
+        self.stack.borrow().last().unwrap().clone()
+    }
+
+    /// The properties of the parent node, or `None` if there is no parent.
+    fn parent_properties(&self) -> Option<Properties> {
+        let borrow = self.stack.borrow();
+        if borrow.len() >= 2 {
+            Some(borrow[borrow.len() - 2].clone())
+        } else {
+            None
+        }
+    }
+}
+impl Visitor<(), PropertiesGuard> for InjectVisitor {
+    fn enter_literal<'a>(&mut self, path: &WalkPath, node: &mut ViewMutLiteral<'a>) -> EnterResult {
+        let _guard = if let ViewMutLiteral::LiteralInfinityExpression(_) = *node {
+            // In JavaScript, this is not considered pure data as there is no syntax for it.
+            PropertiesGuard::complex(self, path)
+        } else {
+            // Other literals are pure data.
+            PropertiesGuard::pure_data(self, 1, path)
+        };
+        // No need to visit the children.
+        Ok(VisitMe::DoneHere)
+    }
+
+    fn enter_spread_element_or_expression<'a>(
+        &mut self,
+        path: &WalkPath,
+        _node: &mut ViewMutSpreadElementOrExpression<'a>,
+    ) -> EnterResult {
+        // Pure data so far.
+        return Ok(VisitMe::HoldThis(PropertiesGuard::pure_data(self, 1, path)));
+    }
+
+    fn enter_object_expression(
+        &mut self,
+        path: &WalkPath,
+        _node: &mut ObjectExpression,
+    ) -> EnterResult {
+        // Pure data so far.
+        return Ok(VisitMe::HoldThis(PropertiesGuard::pure_data(self, 1, path)));
```

Comment:
shouldn't we handle each `ObjectProperty` items ?
are ShorthandProperty and ShorthandProperty pure data?

Reply:
Good catch for `ObjectProperty`.